## **LAB-3**
### **Name: Anjali Nagpal Lab 3**
### **StudenID: 8895806**

## PART-A

#### Problem Statement: Consider the dataset Credit Card Fraud Detection from Kaggle and build a machine-learning model that detects whether a credit card transaction is fraudulent. Demonstrate the steps of data preprocessing and analysis, consider applying train (0.7) and test (0.3), using the logistic regression to build the model, and evaluate to determine the accuracy.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, r2_score, mean_squared_error, mean_absolute_error
from sklearn.preprocessing import StandardScaler, LabelEncoder

#### Load the dataset

In [2]:
data = pd.read_csv('C:/Users/Arin/Downloads/creditcarddata/creditcard.csv')
data.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

#### Explore the dataset

In [3]:
print(data.shape)
print(data.info())
print(data['Class'].value_counts())

(284807, 31)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 2

#### Data preprocessing

In [4]:
# Checking Duplicate rows
print("Number of duplicate rows:", data.duplicated().sum())

# Remove duplicate rows
data.drop_duplicates(inplace=True)

# Reset the index after removing duplicates
data.reset_index(drop=True, inplace=True)

print("Number of duplicate rows after cleaning:", data.duplicated().sum())

# Handle missing values (if any)
data.dropna(inplace=True)

print('Number of normal transaction', len(data[data.Class == 0]))
print('Number of fraud transaction', len(data[data.Class == 1]))

# Separate features (X) and target variable (y)
X = data.drop('Class', axis=1)
y = data['Class']

# Scale the features (except the 'Class' column)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


Number of duplicate rows: 1081
Number of duplicate rows after cleaning: 0
Number of normal transaction 283253
Number of fraud transaction 473


#### Split the dataset into a training set (70%) and a testing set (30%)

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.3, random_state=42)


#### Train the model

In [6]:
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

LogisticRegression()

#### Predict the model

In [7]:
y_pred = log_reg.predict(X_test)

#### Evaluate the model

In [8]:
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:\n", conf_matrix)
print("Classification Report:\n", classification_rep)

Accuracy: 0.9991658638595832
Confusion Matrix:
 [[84973    11]
 [   60    74]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     84984
           1       0.87      0.55      0.68       134

    accuracy                           1.00     85118
   macro avg       0.93      0.78      0.84     85118
weighted avg       1.00      1.00      1.00     85118



#### Conclusion

##### The model achieves a very high accuracy of 99.92%. However, when dealing with imbalanced datasets like credit card fraud detection, where non-fraudulent transactions outnumber fraudulent ones, accuracy alone may not be sufficient.

##### The confusion matrix sheds light on the types of errors the model is making. Specifically, there were 60 fraudulent transactions classified as non-fraudulent (false negatives), and 11 non-fraudulent transactions classified as fraudulent (false positives).

## PART-B

#### Problem Statement: Use the following insurance dataset and build a predictive system to predict insurance costs. Demonstrate the steps of data preprocessing and analysis, consider applying train (0.7) and test (0.3), using linear regression to build the model, and evaluate the accuracy of predicting the insurance cost.

#### Load the dataset

In [9]:
# Load the dataset
insurance_data = pd.read_csv('C:/Users/Arin/Downloads/creditcarddata/insurance.csv')
insurance_data.head(10)


age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520
5   31  female  25.740         0     no  southeast   3756.62160
6   46  female  33.440         1     no  southeast   8240.58960
7   37  female  27.740         3     no  northwest   7281.50560
8   37    male  29.830         2     no  northeast   6406.41070
9   60  female  25.840         0     no  northwest  28923.13692

#### Explore the dataset

In [10]:
print(insurance_data.shape)
print(insurance_data.info())
print(insurance_data.describe())


(1338, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB
None
               age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.

#### Data preprocessing

In [11]:
print("Number of duplicate rows:", insurance_data.duplicated().sum())

# Remove duplicate rows
insurance_data.drop_duplicates(inplace=True)

# Reset the index after removing duplicates
insurance_data.reset_index(drop=True, inplace=True)

print("Number of duplicate rows after cleaning:",
      insurance_data.duplicated().sum())

# Handle missing values (if any)
insurance_data.dropna(inplace=True)

# Convert categorical variables into numerical using one-hot encoding (if necessary)

categorical_columns = ['sex', 'smoker','region']
# Now changing catergorical variables into numeric variables
label_encoder = LabelEncoder()
for cat_col in categorical_columns:
    insurance_data[cat_col] = label_encoder.fit_transform(
        insurance_data[cat_col])

display(insurance_data)

# Separate features (X) and target variable (y)
X = insurance_data.drop('charges', axis=1)
y = insurance_data['charges']


Number of duplicate rows: 1
Number of duplicate rows after cleaning: 0


age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       3  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       1  21984.47061
4      32    1  28.880         0       0       1   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1332   50    1  30.970         3       0       1  10600.54830
1333   18    0  31.920         0       0       0   2205.98080
1334   18    0  36.850         0       0       2   1629.83350
1335   21    0  25.800         0       0       3   2007.94500
1336   61    0  29.070         0       1       1  29141.36030

[1337 rows x 7 columns]

#### Split the dataset into a training set (70%) and a testing set (30%)

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42)


#### Train the model

In [13]:
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

#### Predict the model

In [14]:
y_pred = lin_reg.predict(X_test)

#### Evaluate the model

In [15]:
accuracy = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
# Print the evaluation metrics
print(f"Mean Squared Error (MSE): {mse}")
print(f"Mean Absolute Error (MAE): {mae}")
print(f"R-squared (R2) Score: {accuracy}")

Mean Squared Error (MSE): 38935221.010640755
Mean Absolute Error (MAE): 4182.803777070157
R-squared (R2) Score: 0.7724652729621758


#### Conclusion
##### A lower MSE value indicates that the model's predictions are closer to the actual values.
##### A higher R-squared value indicates a better fit of the model to the data and better predictive performance.